astra/src/main/java/com/slack/astra/zipkinApi/TraceFetcher.java

```diff
@@ -0,0 +1,372 @@
+package com.slack.astra.zipkinApi;
+
+import static com.slack.astra.metadata.dataset.DatasetPartitionMetadata.MATCH_ALL_DATASET;
+
+import brave.Tracing;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.MapperFeature;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.json.JsonMapper;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.protobuf.ByteString;
+import com.slack.astra.blobfs.BlobStore;
+import com.slack.astra.logstore.LogMessage;
+import com.slack.astra.logstore.LogWireMessage;
+import com.slack.astra.proto.service.AstraSearch;
+import com.slack.astra.server.AstraQueryServiceBase;
+import com.slack.astra.util.JsonUtil;
+import java.io.IOException;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+import org.json.JSONObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TraceFetcher {
+  private static final Logger LOG = LoggerFactory.getLogger(TraceFetcher.class);
+
+  private final AstraQueryServiceBase searcher;
+  private final BlobStore blobStore;
+
+  private final int defaultMaxSpans;
+  private final int defaultLookbackMins;
+  private final long defaultDataFreshnessInSeconds;
+
+  public static final String TRACE_CACHE_PREFIX = "traceCacheData";
+
+  public TraceFetcher(
+      AstraQueryServiceBase searcher,
+      BlobStore blobStore,
+      int defaultMaxSpans,
+      int defaultLookbackMins,
+      long defaultDataFreshnessInSeconds) {
+    this.searcher = searcher;
+    this.blobStore = blobStore;
+    this.defaultMaxSpans = defaultMaxSpans;
+    this.defaultLookbackMins = defaultLookbackMins;
+    this.defaultDataFreshnessInSeconds = defaultDataFreshnessInSeconds;
+  }
+
+  private static class Result {
+    final String rawJson;
+    final List<ZipkinSpanResponse> spans;
+
+    Result(String rawJson) {
+      this.rawJson = rawJson;
+      this.spans = null;
+    }
+
+    Result(List<ZipkinSpanResponse> spans) {
+      this.spans = spans;
+      this.rawJson = null;
+    }
+
+    boolean isCached() {
+      return rawJson != null;
+    }
+  }
```

Comment:
could be a record class

```suggestion
  private static record Result(String rawJson, List<ZipkinSpanResponse> {
    Result(String rawJson) {
      this.rawJson = rawJson;
      this.spans = null;
    }

    Result(List<ZipkinSpanResponse> spans) {
      this.spans = spans;
      this.rawJson = null;
    }

    boolean isCached() {
      return rawJson != null;
    }
  }
```